File: aoc/2022/aoc-dart-2022/lib/day_1.dart
import 'package:aoc_dart_2022/common/common.dart';

int partOne (String inputFilepath) {
    List<String> lines = read_in_file(inputFilepath);
    
    List<int> sums = [];

    int currSum = 0;
    for (var line in lines) {
      if (line == "") {
        sums.add(currSum);
        currSum = 0;
      } else {
        currSum += int.parse(line);
      }
    }

    sums.add(currSum);

    sums.sort();

    return sums.last;
}

int partTwo (String inputFilepath) {

    List<String> lines = read_in_file(inputFilepath);
    
    List<int> sums = [];

    int currSum = 0;
    for (var line in lines) {
      if (line == "") {
        sums.add(currSum);
        currSum = 0;
      } else {
        currSum += int.parse(line);
      }
    }

    sums.add(currSum);

    sums.sort();

    return sums[sums.length-1] + sums[sums.length-2] + sums[sums.length-3]; 
}

File: aoc/2022/aoc-rust-2022/src/day_1.rs
use std::fs::File;
use std::io::{self, BufRead};

pub fn part_one(filepath: String) -> u32 {

    let input_file = File::open(filepath).expect("File not found!");
    let reader = io::BufReader::new(input_file);

    let mut curr_sum : u32 = 0;
    let mut cal_sums : Vec<u32> = vec![];

    for line in reader.lines() {
        match line {
            Ok(line) => {
                match line.as_str() {
                    "" => {
                        cal_sums.push(curr_sum);
                        curr_sum = 0;
                    }
                    _ => {
                        curr_sum += line.parse::<u32>().unwrap();
                    }
                }
                
            }

            Err(error) => {
                eprintln!("Error reading line: {}", error);
            }
        }
    }

    cal_sums.push(curr_sum);
    cal_sums.sort();
    
    return *cal_sums.last().unwrap();

}

pub fn part_two(filepath: String) -> u32 {

    let input_file = File::open(filepath).expect("File not found!");
    let reader = io::BufReader::new(input_file);

    let mut curr_sum : u32 = 0;
    let mut cal_sums : Vec<u32> = vec![];

    for line in reader.lines() {
        match line {
            Ok(line) => {
                match line.as_str() {
                    "" => {
                        cal_sums.push(curr_sum);
                        curr_sum = 0;
                    }
                    _ => {
                        curr_sum += line.parse::<u32>().unwrap();
                    }
                }
                
            }

            Err(error) => {
                eprintln!("Error reading line: {}", error);
            }
        }
    }

    cal_sums.push(curr_sum);
    cal_sums.sort();

    let top_three : u32 = cal_sums[cal_sums.len()-1] +
                         cal_sums[cal_sums.len()-2] +
                         cal_sums[cal_sums.len()-3] ;

    return top_three;    

}
